guillotina/asgi.py

```diff
@@ -0,0 +1,208 @@
+from aiohttp.streams import EmptyStreamReader
+from guillotina import glogging
+from guillotina import task_vars
+from guillotina.exceptions import ConflictError
+from guillotina.exceptions import TIDConflictError
+from guillotina.request import Request
+from typing import Optional
+
+import asyncio
+
+
+logger = glogging.getLogger('guillotina')
+
+
+def headers_to_list(headers):
+    return [[k.encode(), v.encode()] for k, v in headers.items()]
+
+
+class AsgiStreamReader(EmptyStreamReader):
+    """Dummy implementation of StreamReader"""
+
+    def __init__(self, receive):
+        self.finished = False
+        self._asgi_receive = receive
+        self._buffer = b""
+
+    async def readany(self):
+        from guillotina.files import CHUNK_SIZE
+        return await self.read(CHUNK_SIZE)
+
+    async def readexactly(self, n: int) -> bytes:
+        data = await self.read(n)
+        if len(data) < n:
+            raise asyncio.IncompleteReadError(data, n)
+        return data
+
+    async def read(self, n: int = -1):
+        if self._buffer:
+            data = self._buffer
+        else:
+            data = b""
+
+        while (n == -1 or len(data) < n) and not self.finished:
+            chunk = await self.receive()
+            data += chunk
+
+        if n == -1:
+            self._buffer = None
+            return data
+        else:
+            req_chunk, rest = data[:n], data[n:]
+            self._buffer = rest
+            return req_chunk
+
+    async def receive(self):
+        payload = await self._asgi_receive()
+        self.finished = not payload.get("more_body", False)
+        return payload["body"]
+
+
+class AsgiStreamWriter():
```

Comment:
@dmanchon what do you think moving the writer to `response.py`?

Reply:
good idea!